File: application/test/functional/gateway/OutboundEventsFlowIT.scala
package functional.gateway

import java.util.UUID
import java.util.concurrent.TimeUnit

import common.FunctionalSpec
import common.api.{ConnectionsClient, EndpointsClient}
import fixtures.EndpointFixtures
import play.api.libs.json.Json
import play.api.test.Helpers._


class OutboundEventsFlowIT extends FunctionalSpec with EndpointsClient with ConnectionsClient {
  private val DEFAULT_TIMEOUT = 5
  private val path = "/outbound-flow"

  override protected def beforeAll(): Unit = {
    createAndAssert(EndpointFixtures.fromPath(path))
  }

  feature("WS connection receive events sent by backends through http endpoint") {
    scenario("WS connection receive outbound events") {

      Given("a ws connection")
      val connectionId = UUID.randomUUID().toString
      val receivedMessages = await(wsClient.connect(wsConnectionUrl(path, connectionId)))._1

      When("http request is made to connections api")
      val payload = Json.stringify(Json.obj("foo" -> "bar"))
      sendEvent(connectionId, payload)

      Then("message received by ws connection")
      receivedMessages.poll(DEFAULT_TIMEOUT, TimeUnit.SECONDS) mustBe payload
    }

    scenario("WS connection receive multiple outbound events in the same order") {

      Given("a ws connection")
      val connectionId = UUID.randomUUID().toString
      val receivedMessages = await(wsClient.connect(wsConnectionUrl(path, connectionId)))._1

      When("send multiple http requests to connwections api")
      val payload1 = Json.stringify(Json.obj("foo" -> "bar"))
      val payload2 = Json.stringify(Json.obj("a" -> "b"))
      val payload3 = Json.stringify(Json.obj("id" -> 1, "msg" -> "some msg"))

      sendEvent(connectionId, payload1)
      sendEvent(connectionId, payload2)
      sendEvent(connectionId, payload3)

      Then("message received by ws connection")
      receivedMessages.poll(DEFAULT_TIMEOUT, TimeUnit.SECONDS) mustBe payload1
      receivedMessages.poll(DEFAULT_TIMEOUT, TimeUnit.SECONDS) mustBe payload2
      receivedMessages.poll(DEFAULT_TIMEOUT, TimeUnit.SECONDS) mustBe payload3
    }
  }
}

File: ws-gateway-functional-tests/src/test/java/com/cosmin/wsgateway/tests/gateway/OutboundEventsFlowIT.java
package com.cosmin.wsgateway.tests.gateway;

import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.cosmin.wsgateway.tests.BaseTestIT;
import com.cosmin.wsgateway.tests.Tags;
import com.cosmin.wsgateway.tests.common.EndpointFixtures;
import com.cosmin.wsgateway.tests.common.JsonUtils;
import com.cosmin.wsgateway.tests.utils.Conditions;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.awaitility.Awaitility;
import org.junit.jupiter.api.Test;

@Tags.Gateway
@Tags.PubSub
public class OutboundEventsFlowIT extends BaseTestIT {
    @Test
    public void gatewayClientReceivesEventsSentViaConnectionsApi() {
        Given("an websocket connection");
        var connectionId = UUID.randomUUID().toString();
        var event1 = JsonUtils.toJson(Map.of("foo", "bar"));
        var event2 = JsonUtils.toJson(List.of("foo", "bar2"));

        var endpoint = EndpointFixtures.getRepresentation("/outbound-events/test-1");
        endpointsClient.createAndAssert(endpoint);
        var connection = webSocketClient.connect(endpoint.getPath(), connectionId);
        Awaitility.await().during(Duration.ofMillis(200)).until(() -> true);

        When("Connections outbound api is called");
        connectionsClient.sendEvent(connectionId, event1);
        connectionsClient.sendEvent(connectionId, event2);

        Then("Gateway client receives msgs via api");
        Awaitility.await("should receive 2 messages")
                .atMost(Duration.ofSeconds(20))
                .until(Conditions.receivedMessages(connection), is(2));

        assertTrue(connection.getReceivedMessages().contains(event1));
        assertTrue(connection.getReceivedMessages().contains(event2));
    }
}
